### ISSUE TITLE
Random results

### ISSUE THREAD

**Issue by nikita250996** [skorch-dev/skorch#522]
Hello. I've been trying to use skorch properly, but noticed that without changing code at all I'm getting different results for the same input.
What is wrong?
[strange.zip](https://github.com/skorch-dev/skorch/files/3594506/strange.zip)

**Comment by BenjaminBossan** [comment]
I haven't looked at the code, but just for starters: Did you fix all the random seeds (numpy, torch, random_state in sklearn etc.)? Does the randomness persist if you only use CPU and not GPU?

**Comment by nikita250996** [comment]
> I haven't looked at the code, but just for starters: Did you fix all the random seeds (numpy, torch, random_state in sklearn etc.)? Does the randomness persist if you only use CPU and not GPU?

Well, as you suggested, I tried CPU only, randomness is still here.
My code for seed looks like this.
`random.seed(0)
np.random.seed(0)
torch.manual_seed(0)
torch.cuda.manual_seed(0)
torch.backends.cudnn.deterministic = True`

**Comment by BenjaminBossan** [comment]
I don't really see anything wrong in the code. Could you try the following and check if anything changes:

```python
net = NeuralNetClassifier(
    PretrainedModel, 
    criterion=torch.nn.CrossEntropyLoss,
    lr=0.01,
    batch_size=16,
    max_epochs=5,
    module__output_features=2,
    optimizer=torch.optim.SGD,
    optimizer__momentum=0.85,
    optimizer__nesterov=False,
    # iterator_train__shuffle=True,
    iterator_train__num_workers=1,
    iterator_valid__shuffle=False,
    iterator_valid__num_workers=1,
    train_split=False,
    # callbacks=[lrscheduler, checkpoint],
    device='cuda'
)
```

Also, could you try to turn off the random transformers in your datasets (such as `RandomResizedCrop`).

**Comment by nikita250996** [comment]
I used your code and received an error about best_model.pt, so I also commented this line
`net.load_params(f_params='best_model.pt')`
It killed both randomness and classification, now it thinks all pictures belongs to one class.
By the way, I also wanted to say that sometimes another problem appears. With any code the net may classify every picture to be a part of just one class. It only appears with skorch.
When transformation code looks like this
`train_transforms = transforms.Compose([
    transforms.Resize((224, 224)),#0.89112/0.82795
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
])`
randomness is present.
Turning off all random transofrms and editing the net in the way you suggested gives the same result of lacking at least one picture belongins to a second class.

**Comment by BenjaminBossan** [comment]
> now it thinks all pictures belongs to one class.

So I think what happens is that the random transforms are not covered by your seeds. I'm not sure if there is anything that can be done. Turning off the random transforms, however, leads to worse training, resulting in the net not learning properly.

You could try to undo some of the changes I suggested above to see if that still leads to randomness or not. Otherwise, I can't really help you, there is not much in skorch that would introduce randomness (only the train split comes to mind).

**Comment by nikita250996** [comment]
May I also ask one more question about skorch in this thread? It's not an issue but the thing I don't understand how to implement.

**Comment by BenjaminBossan** [comment]
Feel free to open a new issue, we have no limit :)

**Comment by nikita250996** [comment]
Thank you.
